### PR TITLE
callback helper and utility methods

### PR DESCRIPTION
**Pull request by jdhudson3**

### CODE REVIEW

File: src/main/java/com/bandwidth/sdk/messaging/models/Message.java [hunk 1]
```diff
@@ -36,4 +37,14 @@ public boolean isMms(){
         return !isSms();
     }
 
+    @JsonIgnore
+    public Set<String> getReplyNumbers(){
+        Set<String> toNumbers = new TreeSet<>();
+        getTo()
+            .stream()
+            .filter(number -> !getOwner().equals(number))
+            .forEach(number -> toNumbers.add(number));
```
Comment (fuchsnj) [inline src/main/java/com/bandwidth/sdk/messaging/models/Message.java]: You can use `.collect(...)` to put all the elements in a collection

Comment (fuchsnj) [reply src/main/java/com/bandwidth/sdk/messaging/models/Message.java]: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#collect-java.util.stream.Collector-

File: src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java [hunk 2]
```diff
@@ -18,6 +20,17 @@
...
     public abstract Optional<String> getTo();
 
+
+    @JsonIgnore
+    public boolean isIncomingMessage(){
+        return getTo().isPresent();
+    }
+
+    @JsonIgnore
+    public boolean isDLR(){
```
Comment (fuchsnj) [inline src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java]: I think we should try not using acronyms here. Most people outside of the industry probably don't know what it is.
`isDeliveryReceipt()`?

File: src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java [hunk 3]
```diff
@@ -18,6 +20,17 @@
     public abstract Optional<Integer> getErrorCode();
     public abstract Optional<String> getTo();
 
+
+    @JsonIgnore
+    public boolean isIncomingMessage(){
+        return getTo().isPresent();
```
Comment (fuchsnj) [inline src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java]: All messages have "to" numbers, this doesn't determine if it's incoming. You should look at the event type instead

File: src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java [hunk 4]
```diff
@@ -28,4 +41,28 @@ public boolean isMms(){
...
+    @JsonIgnore
+    public boolean isError(){
+        return getErrorCode().isPresent();
+    }
+
+    @JsonIgnore
+    public Optional<MessageErrorTypes> getErrorType(){
+        Optional<MessageErrorTypes> errorType = Optional.empty();
+
+        if (getErrorCode().isPresent()){
```
Comment (fuchsnj) [inline src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java]: You can use `.map()` to simplify the logic here
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

File: src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java (hunk 3 above)
Comment (jmulford-bandwidth) [reply src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java]: I think the logic here is that the callback won't have a "to" field if the callback isn't an incoming message. I'm not sure if that is the case 100% of the time. All callbacks have a "message" field which has a "message.to" field, which is different than the callback's "to" field.

Not saying I disagree with looking at the event type though

Comment (fuchsnj) [reply src/main/java/com/bandwidth/sdk/messaging/models/MessageEvent.java]: @jmulford-bandwidth you are correct, I thought this was on the message object itself, but this is the message event. It's still probably better to look at the event type, there may be cases where the "to" field could exist for other types (potentially in the future). We would just have a constant "message-received" and check against that

**Pull request merged**